File: src/androidTest/java/com/todoroo/astrid/gtasks/GtasksListServiceTest.java
package com.todoroo.astrid.gtasks;

import com.google.api.client.util.DateTime;
import com.google.api.services.tasks.model.TaskList;
import com.google.api.services.tasks.model.TaskLists;
import com.todoroo.astrid.dao.Database;
import com.todoroo.astrid.dao.StoreObjectDao;
import com.todoroo.astrid.test.DatabaseTestCase;

import org.tasks.makers.RemoteGtaskListMaker;

import javax.inject.Inject;

import static com.natpryce.makeiteasy.MakeItEasy.with;
import static java.util.Arrays.asList;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.tasks.time.DateTimeUtils.currentTimeMillis;
import static org.tasks.makers.GtaskListMaker.LAST_SYNC;
import static org.tasks.makers.GtaskListMaker.NAME;
import static org.tasks.makers.GtaskListMaker.REMOTE_ID;
import static org.tasks.makers.GtaskListMaker.newGtaskList;
import static org.tasks.makers.RemoteGtaskListMaker.newRemoteList;

public class GtasksListServiceTest extends DatabaseTestCase {

    @Inject Database database;

    private StoreObjectDao storeObjectDao;
    private GtasksListService gtasksListService;

    @Override
    public void setUp() {
        super.setUp();
        storeObjectDao = spy(new StoreObjectDao(database));
        gtasksListService = new GtasksListService(storeObjectDao);
    }

    public void testCreateNewList() {
        setLists(newRemoteList(
                with(RemoteGtaskListMaker.REMOTE_ID, "1"),
                with(RemoteGtaskListMaker.NAME, "Default")));

        verify(storeObjectDao).persist(newGtaskList(
                with(REMOTE_ID, "1"),
                with(NAME, "Default")));
    }

    public void testGetListByRemoteId() {
        GtasksList list = newGtaskList(with(REMOTE_ID, "1"));
        storeObjectDao.createNew(list.getStoreObject());

        assertEquals(list, gtasksListService.getList("1"));
    }

    public void testGetListReturnsNullWhenNotFound() {
        assertNull(gtasksListService.getList("1"));
    }

    public void testDeleteMissingList() {
        storeObjectDao.createNew(newGtaskList(with(REMOTE_ID, "1")).getStoreObject());

        setLists(newRemoteList(with(RemoteGtaskListMaker.REMOTE_ID, "2")));

        verify(storeObjectDao).delete(1L);
    }

    public void testUpdateListName() {
        storeObjectDao.createNew(newGtaskList(
                with(REMOTE_ID, "1"),
                with(NAME, "oldName")).getStoreObject());

        setLists(newRemoteList(
                with(RemoteGtaskListMaker.REMOTE_ID, "1"),
                with(RemoteGtaskListMaker.NAME, "newName")));

        assertEquals("newName", storeObjectDao.getGtasksList(1).getName());
    }

    public void testNewListLastSyncIsZero() {
        setLists(new TaskList().setId("1"));

        assertEquals(0L, gtasksListService.getList("1").getLastSync());
    }

    public void testNewListNeedsUpdate() {
        TaskList taskList = new TaskList().setId("1").setTitle("Default").setUpdated(new DateTime(currentTimeMillis()));

        setLists(taskList);

        assertEquals(
                asList(newGtaskList(with(REMOTE_ID, "1"), with(LAST_SYNC, 0L))),
                gtasksListService.getListsToUpdate(new TaskLists().setItems(asList(taskList))));
    }

    private void setLists(TaskList... list) {
        gtasksListService.updateLists(new TaskLists().setItems(
                asList(list)));
    }
}

File: app/src/androidTest/java/com/todoroo/astrid/gtasks/GtasksListServiceTest.kt
package com.todoroo.astrid.gtasks

import com.google.api.services.tasks.model.TaskList
import com.natpryce.makeiteasy.MakeItEasy.with
import com.todoroo.astrid.service.TaskDeleter
import dagger.hilt.android.testing.HiltAndroidTest
import dagger.hilt.android.testing.UninstallModules
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.tasks.LocalBroadcastManager
import org.tasks.data.dao.CaldavDao
import org.tasks.data.entity.CaldavAccount
import org.tasks.data.entity.CaldavCalendar
import org.tasks.injection.InjectingTestCase
import org.tasks.injection.ProductionModule
import org.tasks.makers.RemoteGtaskListMaker
import org.tasks.makers.RemoteGtaskListMaker.newRemoteList
import javax.inject.Inject

@UninstallModules(ProductionModule::class)
@HiltAndroidTest
class GtasksListServiceTest : InjectingTestCase() {
    @Inject lateinit var taskDeleter: TaskDeleter
    @Inject lateinit var localBroadcastManager: LocalBroadcastManager
    @Inject lateinit var caldavDao: CaldavDao

    private lateinit var gtasksListService: GtasksListService

    @Before
    override fun setUp() {
        super.setUp()
        gtasksListService = GtasksListService(caldavDao, taskDeleter, localBroadcastManager)
    }

    @Test
    fun testCreateNewList() = runBlocking {
        setLists(
                newRemoteList(
                        with(RemoteGtaskListMaker.REMOTE_ID, "1"), with(RemoteGtaskListMaker.NAME, "Default")))
        assertEquals(
            CaldavCalendar(id = 1, account = "account", uuid = "1", name = "Default"),
            caldavDao.getCalendarById(1L)
        )
    }

    @Test
    fun testGetListByRemoteId() = runBlocking {
        val list = CaldavCalendar(uuid = "1")
        caldavDao.insert(list)
        assertEquals(list, caldavDao.getCalendarByUuid("1"))
    }

    @Test
    fun testGetListReturnsNullWhenNotFound() = runBlocking {
        assertNull(caldavDao.getCalendarByUuid("1"))
    }

    @Test
    fun testDeleteMissingList() = runBlocking {
        caldavDao.insert(CaldavCalendar(account = "account", uuid = "1"))
        val taskList = newRemoteList(with(RemoteGtaskListMaker.REMOTE_ID, "2"))
        setLists(taskList)
        assertEquals(
            listOf(CaldavCalendar(id = 2, account = "account", uuid = "2", name = "Default")),
            caldavDao.getCalendarsByAccount("account")
        )
    }

    @Test
    fun testUpdateListName() = runBlocking {
        val calendar = CaldavCalendar(uuid = "1", name = "oldName", account = "account")
        caldavDao.insert(calendar)
        setLists(
                newRemoteList(
                        with(RemoteGtaskListMaker.REMOTE_ID, "1"), with(RemoteGtaskListMaker.NAME, "newName")))
        assertEquals("newName", caldavDao.getCalendarById(calendar.id)!!.name)
    }

    @Test
    fun testNewListLastSyncIsZero() = runBlocking {
        setLists(TaskList().setId("1"))
        assertEquals(0L, caldavDao.getCalendarByUuid("1")!!.lastSync)
    }

    private suspend fun setLists(vararg list: TaskList) {
        val account = CaldavAccount(
            username = "account",
            uuid = "account",
        )
        caldavDao.insert(account)
        gtasksListService.updateLists(account, listOf(*list))
    }
}